### DISCUSSION TITLE
Current Model Structure And Potential Improvements

### DISCUSSION THREAD

**Discussion by TimothyWillard** [ACCIDDA/vaxflux#1] [Ideas]
# The Current Model

Let $i = 1, \dots, N_\mathrm{season}$, $j = 1, \dots, N_\mathrm{strata}$, $l = 1, \dots, N_\mathrm{region}$. The observed uptake at time $t$ into season $i$ for strata $j$ and region $l$ is given by $y_{ijl}(t)$. Then,

$$ k_i \sim \mathcal{N}\left( \mu_k, \sigma_k^2 \right) $$

$$ r_i \sim \mathcal{N}\left( \mu_r, \sigma_r^2 \right) $$

$$ s_i \sim \mathcal{N}\left( \mu_s, \sigma_s^2 \right) $$

$$ \delta k_j \sim \mathcal{N}\left( \mu_{\delta k_\mathrm{strata}}, \sigma_{\delta k_\mathrm{strata}}^2 \right) $$

$$ \delta r_j \sim \mathcal{N}\left( \mu_{\delta r_\mathrm{strata}}, \sigma_{\delta r_\mathrm{strata}}^2 \right) $$

$$ \delta s_j \sim \mathcal{N}\left( \mu_{\delta s_\mathrm{strata}}, \sigma_{\delta s_\mathrm{strata}}^2 \right) $$

$$ \delta k_l \sim \mathcal{N}\left( \mu_{\delta k_\mathrm{region}}, \sigma_{\delta k_\mathrm{region}}^2 \right) $$

$$ \delta r_l \sim \mathcal{N}\left( \mu_{\delta r_\mathrm{region}}, \sigma_{\delta r_\mathrm{region}}^2 \right) $$

$$ \delta s_l \sim \mathcal{N}\left( \mu_{\delta s_\mathrm{region}}, \sigma_{\delta s_\mathrm{region}}^2 \right) $$

$$ \epsilon \sim \mathrm{HalfNormal}\left( \sigma_{\epsilon} \right) $$

$$ y_{ijl}\left(t\right) \sim \mathcal{N}\left( f\left( t \vert K_{ijl}, R_{ijl}, S_{ijl} \right), \epsilon^2 \right) $$

where $K_{ijl} = k_i + \delta k_j + \delta k_l$, $R_{ijl} = r_i + \delta r_j + \delta r_l$, $S_{ijl} = s_i + \delta s_j + \delta s_l$, and $f\left( t \vert k, r, s \right) = \mathrm{expit}(k)\mathrm{expit}(r(t-s))$. The model can be simplified by setting $i$, $j$, and/or $l$ to 1 for complete pooling across seasons, strata, and/or regions. Furthermore, the strata/region specific factors can be removed by setting them to 0 rather than drawing from a normally distributed prior.

A nice upside to the current model are:

1. The max uptake is reasonably bounded, $0 < K_{ijl} < 1$.
2. It's a very interpretable model: a) $\mathrm{expit}\left(K_{ijl}\right)$ corresponds to the max uptake, b) $s$ corresponds to the halfway point in a campaign, $f(s \vert k, r, s) = \mathrm{expit}(k)/2$, and c) the $r$ can be understood as $f(s \vert k, r, s) \approx (1/2) + (r/4)x$ summarized as "at the peak of the flu vaccine uptake the vaccination rate for the population is $r/4$" measured in percent per a day.

# Potential Issues

There are a few issues with this current model ranging in severity.

1. This model seems to *overestimate* uptake when $t < s$, is there an easy way that we could make $f$ asymmetric about $s$ without introducing too many parameters?
3. The priors for $k$ allow for the max uptake to be less than the observed max, is there any easy way to specify the priors such that this cannot happen? This might be more of an implementation/coding issue then a model issue per se.
4. The current observational model is normally distributed about $f$, which can allow for physically impossible uptake values such as less than 0. Is there a way to bound the error such that these unrealistic values cannot be produced?
5. Especially pedantic, but I think $k$ should be changed to $m$. The other parameters correspond to the name well, $r$ for "rate" and $s$ for "switch point". Also, this would allow for $i$, $j$, $k$ for indexing.
6. Is it worth trying to consolidate both strata and region into a shared "factor" concept? At the moment these portions of the model are exchangeable (i.e. you could provide age groups for region and states for strata and it wouldn't change anything). Again, this might be more of an implementation issue.

**Comment by jc-macdonald** [comment]
regarding 3.  I would suggest switching the priors to log normal, gamma, or beta - this will force all parameters to have positive values (since the support for these distributions for a random variable $X$ is $x > 0$, and for the beta distribution $0 < x < 1$.  This change in prior would also relax the assumption that the distribution is symmetric about the mean.

Regarding 1 & 2 I think it is important to remember that this is a model of the data and no model will perfectly replicate empirical data.  That said, if we were to return to the previous formulation of a logistic model that I was using $\frac{dx}{dt} = rx\left(1-\frac{x}{K}\right),~ x(0) = x_0$ then we can control the maximum (as $K$ is the maximum vaccination coverage).  This model has an exact solution, $x(t) = \frac{x_0Ke^{rt}}{(K-x_0) + x_0e^{rt}}.$

Another possibility is that we can have a function of the type $\frac{K x^{\alpha}}{r + x^{\alpha}} + x_0$ and infer alpha, which will change the inflection point, or as you are calling it, the "switch point" as one varies $\alpha$.  In this formulation the maximum value is $K$ and the "half-saturation" point (in terms of vaccines) is $r$.

**Reply by jc-macdonald** [reply]
To demonstrate the flexibility of the general saturation model here are some different shapes it can have with different parameter values 
![Fig1](https://github.com/user-attachments/assets/230fba3b-ab05-4031-bf19-046ecbda3314)
![Fig2](https://github.com/user-attachments/assets/ec1dd2d2-ca8f-46e7-b5d2-8ae3d657a963)
![Fig3](https://github.com/user-attachments/assets/c5139e69-ea91-4578-a1d7-85aaf785e357)
@TimothyWillard @shauntruelove @twallema

**Reply by twallema** [reply]
Cool, that its flexible indeed! Can you post that figure of your model fit to all those years of data?

**Reply by jc-macdonald** [reply]
> Cool, that its flexible indeed! Can you post that figure of your model fit to all those years of data?

Will do

**Comment by jc-macdonald** [comment]
After discussion with @shauntruelove I'll take a look at the practical identifiability of the more general saturation model, $\frac{Kx^\alpha}{r+x^\alpha} + x_0$.  I'll start with synthetic data and go from there.  If I provide some boiler plate code for identifiability analysis would you be able to incorporate it into the package @TimothyWillard ?  The idea would be that we could have a function the end user could call to test the identifiability of model parameters given the empirical data that they input.

**Reply by TimothyWillard** [reply]
Yeah, definitely! The change would be just swapping out the code here: 

https://github.com/ACCIDDA/vaxflux/blob/41bcf6ed643f09038826091717a4f573d25754f9/src/vaxflux/multilevel_model.py#L317 

if we kept the parameter names the same. Otherwise the work might be a bit more involved but easily doable. I could see a case for wanting to be able to swap out the underlying model from a set of package predefined formulas anyways.

**Reply by twallema** [reply]
- Regarding the observational model (3). From the standpoint of calibrating this model using the vaccine dataset, the observational model is a "weight" function that determines how "bad" a deviation of your model from the data is. F.i. a least-squares cost function would correspond to a normal observational model with epsilon = 1. It is something you ideally choose to be statistically sound, but in practice you may wind up interchanging it to see the result on posterior distributions/goodness-of-fit. Since recording vaccination data are a counting process, I think a poisson or negative binomial observational model should be a statistically sound choice, both distributions can't generate negative uptakes. I think there may be substantial variation in the vaccine uptake year-over-year, so a negative binomial will likely be the better choice.

**Reply by jc-macdonald** [reply]
It might also be worthwhile to have 2-3 models we could ensemble @TimothyWillard (yours, my original model, and the more general saturation model)

**Reply by jc-macdonald** [reply]
> * Regarding the observational model (3). From the standpoint of calibrating this model using the vaccine dataset, the observational model is a "weight" function that determines how "bad" a deviation of your model from the data is. F.i. a least-squares cost function would correspond to a normal observational model with epsilon = 1. It is something you ideally choose to be statistically sound, but in practice you may wind up interchanging it to see the result on posterior distributions/goodness-of-fit. Since recording vaccination data are a counting process, I think a poisson or negative binomial observational model should be a statistically sound choice, both distributions can't generate negative uptakes. I think there may be substantial variation in the vaccine uptake year-over-year, so a negative binomial will likely be the better choice.

I agree, though forcing the underlying parameters to be positive is probably sufficient - if we want to weight one portion of the data over the other this is the correct approach

**Reply by jc-macdonald** [reply]
A note is that at the moment this is a vaccination *prevalence* model, statistically it would be best to fit the *incidence*.  I can write up some code that handles this as a demo if that would be useful
